docker/docker_client.go

```diff
@@ -554,65 +563,72 @@ func (c *dockerClient) getBearerToken(ctx context.Context, challenge challenge,
+		return nil, err
 	}
+	defer resp.Body.Close()
 
-	ping := func(scheme string) error {
-		url := fmt.Sprintf(resolvedPingV2URL, scheme, c.registry)
-		resp, err := c.makeRequestToResolvedURL(ctx, "GET", url, nil, nil, -1, noAuth, nil)
-		if err != nil {
-			logrus.Debugf("Ping %s err %s (%#v)", url, err.Error(), err)
-			return err
-		}
-		defer resp.Body.Close()
-		logrus.Debugf("Ping %s status %d", url, resp.StatusCode)
-		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
-			return errors.Errorf("error pinging registry %s, response code %d (%s)", c.registry, resp.StatusCode, http.StatusText(resp.StatusCode))
-		}
+	logrus.Debugf("Ping %s status %d", url, resp.StatusCode)
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
+		return nil, errors.Errorf("error pinging registry %s, response code %d (%s)",
+			c.registryURL, resp.StatusCode, http.StatusText(resp.StatusCode))
+	}
+
+	return resp, nil
+}
+
+func (c *dockerClient) updateClientByPing(ctx context.Context) error {
+	// Test available schemes for the V2 registry
+	scheme := "https"
+	resp, err := c.ping(ctx, scheme, resolvedPingV2URL)
+	if err != nil && c.skipVerify() {
+		scheme = "http"
+		resp, err = c.ping(ctx, scheme, resolvedPingV2URL)
+	}
+
+	if err == nil {
+		// Update the client if V2 ping worked to make further requests possible
 		c.challenges = parseAuthHeader(resp.Header)
 		c.scheme = scheme
 		c.supportsSignatures = resp.Header.Get("X-Registry-Supports-Signatures") == "1"
-		return nil
-	}
-	err := ping("https")
-	if err != nil && c.insecureSkipTLSVerify {
-		err = ping("http")
-	}
-	if err != nil {
+	} else {
+		// If V2 fails, try V1 if not explicitly disabled
 		err = errors.Wrap(err, "pinging docker registry returned")
 		if c.sys != nil && c.sys.DockerDisableV1Ping {
 			return err
 		}
+
 		// best effort to understand if we're talking to a V1 registry
-		pingV1 := func(scheme string) bool {
-			url := fmt.Sprintf(resolvedPingV1URL, scheme, c.registry)
-			resp, err := c.makeRequestToResolvedURL(ctx, "GET", url, nil, nil, -1, noAuth, nil)
-			if err != nil {
-				logrus.Debugf("Ping %s err %s (%#v)", url, err.Error(), err)
-				return false
-			}
-			defer resp.Body.Close()
-			logrus.Debugf("Ping %s status %d", url, resp.StatusCode)
-			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
-				return false
-			}
-			return true
-		}
-		isV1 := pingV1("https")
-		if !isV1 && c.insecureSkipTLSVerify {
-			isV1 = pingV1("http")
+		_, errV1 := c.ping(ctx, "https", resolvedPingV1URL)
+		if errV1 != nil && c.skipVerify() {
+			_, errV1 = c.ping(ctx, "http", resolvedPingV1URL)
 		}
-		if isV1 {
+		if errV1 == nil {
 			err = ErrV1NotSupported
 		}
 	}
+
 	return err
 }
 
+// detectPropertiesHelper performs the work of detectProperties which executes
+// it at most once.
+func (c *dockerClient) detectPropertiesHelper(ctx context.Context) (err error) {
```

Comment:
(I haven’t reviewed the changes to `detectPropertiesHelper` and its callees at all yet.)

Reply:
@saschagrunert, my brain is already a bit tired but it's not obvious to me why `detectPropertiesHelper` needs to be refactored. Can you elaborate on that?

Reply:
I should not be needed at all, I will revert the changes to keep the code more clean.